основа игры

fedmog1lnkv opened the pull request.

fedmog1lnkv committed:
добавлена основа игры

ArtemBystrovOfficial commented on main.cpp:
```diff
@@ -0,0 +1,26 @@
+﻿#include <fcntl.h>
```
тест

ArtemBystrovOfficial commented on Cell.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "Cell.h"
+
+Cell::Cell() { status = 0; }
```
Cell:Cell() : status(0)
Ты должен инициализировать( вызывать конструкторы других полей прямо из основного конструктора). Это правило Raii

ArtemBystrovOfficial commented on Cell.cpp:
```diff
@@ -0,0 +1,31 @@
+Cell::Cell() { status = 0; }
+
+bool Cell::is_busy() {
+    if (status != 0) {
+        return true;
+    }
+    return false;
```
Урезать до 
return status!=0;

ArtemBystrovOfficial commented on Cell.cpp:
```diff
@@ -0,0 +1,31 @@
+}
+
+ostream& operator<<(ostream& out, const Cell& status_cell) {
+    switch (status_cell.status) {
+    case 0:
+        out << status_cell.empty_cell;
+        break;
+    case 1:
+        out << status_cell.ship_cell;
+        break;
+    case 2:
+        out << status_cell.damaged_ship;
+        break;
+    case 3:
+        out << status_cell.destroyed_ship;
+        break;
+    case 4:
+        out << status_cell.miss_cell;
+        break;
```
На будущее, очень легко со статусами использовать именованное перечисление режимов
Внутри класса пишешь 
`enum class  Status{
   EmptyCell,
   ShipCell... и тд
}
`
После этого ты можешь легко перебирать свитч точно также, только без `магических чисел`
`
func(const Cell::Status & status_cell)
switch(status_ceil.status)
`
А вообще где есть switch, там потенциально мог быть полиморфизм в одну строчку, он для этого и нужен.

ArtemBystrovOfficial commented on Field.cpp:
```diff
@@ -0,0 +1,315 @@
+Field::Field() { fill_map(); }
+
+void Field::fill_map() {
+    for (int i = 0; i < 10; i++) {
+        for (int j = 0; j < 10; j++) {
+            board[i][j] = Cell();
+        }
+    }
```
#define FOR_MAP for(int i=0;i<10; i++) for (int j=0; j<10; j++)

FOR_MAP
     board[i][j] = Cell();

fedmog1lnkv committed:
Улучшения кода

обозначение ячеек - enum
в Player::attack изменена условная конструкция на switch
в классе Player is_bot стало логическим значением

fedmog1lnkv committed:
Фиксы багов, бот

ArtemBystrovOfficial commented on Game.cpp:
```diff
@@ -0,0 +1,217 @@
+#include "Game.h"
+#include "windows.h"
+
+Game::Game() {
+    game_status = 0;
+}
```
: game_status(0)

ArtemBystrovOfficial commented on Player.cpp:
```diff
@@ -0,0 +1,30 @@
+    int status_attack = player_for_attack.map.make_attack(coords);
+
+    switch (status_attack) {
+    case 0:
+        this->radar.change_cell_radar(coords, CSt_t::miss_cell);
+        break;
+    case 1:
+        this->radar.change_cell_radar(coords, CSt_t::damaged_ship);
+        break;
+    case 2:
+        this->radar.mark_destroyed_ship_radar(coords.x, coords.y);
+        player_for_attack.count_ships--;
+    default:
+        break;
```
Оставляем?)

fedmog1lnkv replied:
туда его

fedmog1lnkv committed:
Избавление от "магических чисел"

Вынесение enum class в отдельный файл

fedmog1lnkv commented on Field.cpp:
```diff
@@ -9,422 +9,410 @@ Field::Field() { fill_map(); }
+    }
 }
 
 bool Field::check_ship_placement(int first_x, int first_y, int second_x,
```
поменять на coordinate
также сделать это в функциях:
+ put_ship
+ mark_destroyed_ship_radar

fedmog1lnkv reviewed (commented):
x_first, y_first....
с `int` поменять на структуру `coordinate`

fedmog1lnkv committed:
переписаны функции (аргументы типа coordinate)

изменены функции:
check_ship_placement
put_ship
mark_destroyed_ship_radar

ArtemBystrovOfficial commented on Cell.h:
```diff
@@ -0,0 +1,28 @@
+
+    int weight;
+
+    string empty_cell = "∟ ";
+    string ship_cell = "■ ";
+    string damaged_ship = "□ ";
+    string destroyed_ship = "X ";
+    string miss_cell = "• ";
```
Зачем здесь захватывать пробел если там отступ по умолчанию

ArtemBystrovOfficial commented on EnumClasses.h:
```diff
@@ -0,0 +1,26 @@
+
+// статусы ячейки
+enum class CellStatus {
+    empty_cell = 0,    // 0 - пустая клетка
+    ship_cell = 1,     // 1 - клетка с кораблём
+    damaged_ship = 2,  // 2 - клетка с раненым кораблём
+    destroyed_ship = 3,  // 3 - клетка с уничтоженным кораблём
+    miss_cell = 4  // 4 - клетка с промахом
```
Быстро запутаешься в коде. Поля и переменные имя_переменной, все остальное без черточек. функции camelCase, здесь EmptyCell можно. Чтобы легко можно обращаться было к членам класса можно всем полям прописать m_ (member) и не путать их.
В целом можно пробежаться по https://habr.com/ru/post/480422/. Я не говорю переделывать все я просто для общего развития

ArtemBystrovOfficial commented on Field.h:
```diff
@@ -0,0 +1,87 @@
+
+    // возвращает координаты из строки
+    coordinate get_coords_x_y(string coords);
+
+    // делает из координаты строку
+    string get_coords_string(coordinate cell);
+
+    // проверка выходит ли x и y из поля
+    bool check_area(int x, int y);
+
+    // расстановки кораблей
+    void placement_ships();
+    void auto_placement_ships();
+    void put_ship(coordinate first, coordinate second);
+
+    // проверка можно ли поставить корабль по заданным координатам
+    bool check_ship_placement(coordinate first, coordinate second, int type_ship, bool is_player);
+
+    // возвращает статус атаки
+    StatusAttack make_attack(coordinate coords);
+
+    // возвращает сколько у корабля осталось hp
+    int injure_ship(coordinate coord_ship);
+
+    // отмечает на радаре уничтоженный корабль
+    void mark_destroyed_ship_radar(coordinate coord);
+
+    // меняет на радаре статус клетки
+    void change_cell_radar(coordinate coords, CSt_t new_status);
+
+    // возвращает индекс корабля в массиве ships
+    Ship& find_ship(coordinate coord_cell);
+
+    // расчёт веса клеток на поле
+    void recalculate_weight_board();
+
+    // поиск клетки с макисмальным весом
+    coordinate get_coord_max_weight_cell();
+
+    void show_field();
```
Написал выше, можно менять названия во всех инстанциях сразу одним сочетанием клавиш. Я к тому что очень трудно читать такой объем. 
Цитирую предпочтительный порядок в классе :  типы (включая typedef, using, вложенные классы и структуры), константы, фабричные методы, конструкторы, операторы присваивания, деструкторы, остальные методы, члены данных.

ArtemBystrovOfficial commented on Game.cpp:
```diff
@@ -0,0 +1,219 @@
+        current_player.helper = true;
+    }
+
+    if (next_player.is_bot == true) {
```
хаха,== true ?)

ArtemBystrovOfficial commented on Game.cpp:
```diff
@@ -0,0 +1,219 @@
+        return "Корабль уничтожен";
+    case SA_t::already_use:
+        return "Выстрел по таким координатам уже был";
+    default:
+        return string();
```
Тут нет дефолта

ArtemBystrovOfficial commented on Game.h:
```diff
@@ -0,0 +1,41 @@
+    // определение типа расстановки кораблей и сама расстановка
+    void type_placement_ships();
+
+    void log(string shoot_cell, StatusAttack status_attack);
```
Везде где можно делай const & это очень оптимизирует любое действие. Здесь ты копируешь std::string полностью, ты же помнишь сколько времени занимает новое выделение памяти. Тривиальные типы вроде int можно не брать,  там ссылка даже больше времени займёт, а здесь куча заново будет создаваться и копироваться, так что хорошую привычку можно взять + const & позволяет работать с rvalue. Тоесть ты можешь напрямую туда послать не ссылку а например "text"

ArtemBystrovOfficial commented on Player.cpp:
```diff
@@ -0,0 +1,24 @@
+        this->radar.mark_destroyed_ship_radar(coords);
+        player_for_attack.count_ships--;
+        this->radar.recalculate_weight_board();
+    default:
+        break;
```
Чистим

ArtemBystrovOfficial commented on Ship.cpp:
```diff
@@ -0,0 +1,8 @@
+#include "Ship.h"
+
+Ship::Ship(coordinate coords, int r, int type) {
+    start_coords = coords;
+    rotate = r;
+    hp = type;
+    type_ship = type;
+}
```
: start_coords (), и тд

ArtemBystrovOfficial commented on main.cpp:
```diff
@@ -0,0 +1,23 @@
+    Game game;
+    game.start_preparation();
+
+    // игра между двумя ботами
+    // 1 игра = 3 секунды
+    // для 100 игр = 335,1 секунды
```
В конце почисти не нужные коменты, особенно заметки, это как дополнительно

fedmog1lnkv committed:
Рефракторинг функций
